### PR TITLE
[SPARK-45143][PYTHON][CONNECT] Make PySpark compatible with PyArrow 13.0.0

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1, in PyArrow 13.0.0, the behavior of [Table#to_pandas](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.to_pandas) and [ChunkedArray#to_pandas](https://arrow.apache.org/docs/python/generated/pyarrow.ChunkedArray.html#pyarrow.ChunkedArray.to_pandas) changed, set the `coerce_temporal_nanoseconds=True`

2, there is another undocumented breaking change in data type conversion [`TimestampType#to_pandas_dtype`](https://arrow.apache.org/docs/python/generated/pyarrow.TimestampType.html#pyarrow.TimestampType.to_pandas_dtype):

12.0.1:
```
In [1]: import pyarrow as pa

In [2]: pa.timestamp("us", tz=None).to_pandas_dtype()
Out[2]: dtype('<M8[ns]')

In [3]: pa.timestamp("ns", tz=None).to_pandas_dtype()
Out[3]: dtype('<M8[ns]')

In [4]: pa.timestamp("us", tz="UTC").to_pandas_dtype()
Out[4]: datetime64[ns, UTC]

In [5]: pa.timestamp("ns", tz="UTC").to_pandas_dtype()
Out[5]: datetime64[ns, UTC]
```

13.0.0:
```
In [1]: import pyarrow as pa

In [2]: pa.timestamp("us", tz=None).to_pandas_dtype()
Out[2]: dtype('<M8[us]')

In [3]: pa.timestamp("ns", tz=None).to_pandas_dtype()
Out[3]: dtype('<M8[ns]')

In [4]: pa.timestamp("us", tz="UTC").to_pandas_dtype()
Out[4]: datetime64[us, UTC]

In [5]: pa.timestamp("ns", tz="UTC").to_pandas_dtype()
Out[5]: datetime64[ns, UTC]
```

### Why are the changes needed?
Make PySpark compatible with PyArrow 13.0.0

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
CI

### Was this patch authored or co-authored using generative AI tooling?
NO

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/pandas/typedef/typehints.py [hunk 1]
```diff
@@ -293,7 +293,9 @@ def spark_type_to_pandas_dtype(
         ),
     ):
         return np.dtype("object")
-    elif isinstance(spark_type, types.TimestampType):
+    elif isinstance(spark_type, types.DayTimeIntervalType):
+        return np.dtype("timedelta64[ns]")
+    elif isinstance(spark_type, (types.TimestampType, types.TimestampNTZType)):
         return np.dtype("datetime64[ns]")
     else:
         return np.dtype(to_arrow_type(spark_type).to_pandas_dtype())
```
Comment (zhengruifeng) [inline python/pyspark/pandas/typedef/typehints.py:301]: the behavior of `to_arrow_type(spark_type).to_pandas_dtype()` changed, e.g.:

 `to_arrow_type(DayTimeIntervalType)` -> `pa.timestamp("us", tz="UTC")` -> `datetime64[us, UTC]` in 13.0.0, but `datetime64[ns, UTC]` in 12.0.1

**Review by dongjoon-hyun — COMMENTED** [review]
It's great, @zhengruifeng and @HyukjinKwon .

BTW, how can we solve `mlflow` issue?
```
mlflow 2.6.0 requires pyarrow<13,>=4.0.0, but you have pyarrow 13.0.0 which is incompatible.
```

File: dev/infra/Dockerfile [hunk 2]
```diff
@@ -85,7 +85,7 @@ RUN Rscript -e "devtools::install_version('roxygen2', version='7.2.0', repos='ht
 ENV R_LIBS_SITE "/usr/local/lib/R/site-library:${R_LIBS_SITE}:/usr/lib/R/library"
 
 RUN pypy3 -m pip install numpy 'pandas<=2.0.3' scipy coverage matplotlib
-RUN python3.9 -m pip install numpy 'pyarrow==12.0.1' 'pandas<=2.0.3' scipy unittest-xml-reporting plotly>=4.8 'mlflow>=2.3.1' coverage matplotlib openpyxl 'memory-profiler==0.60.0' 'scikit-learn==1.1.*'
+RUN python3.9 -m pip install numpy pyarrow 'pandas<=2.0.3' scipy unittest-xml-reporting plotly>=4.8 'mlflow>=2.3.1' coverage matplotlib openpyxl 'memory-profiler==0.60.0' 'scikit-learn==1.1.*'
```
Comment (HyukjinKwon) [inline dev/infra/Dockerfile:88]: Let's also upgrade mlflow version.

**Comment by zhengruifeng** [comment]
@dongjoon-hyun I don't see failure in [docker build](https://github.com/zhengruifeng/spark/actions/runs/6184148073/job/16787288022)

```
#21 [15/19] RUN python3.9 -m pip install numpy pyarrow 'pandas<=2.0.3' scipy unittest-xml-reporting plotly>=4.8 'mlflow>=2.3.1' coverage matplotlib openpyxl 'memory-profiler==0.60.0' 'scikit-learn==1.1.*'
#21 43.17 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#21 DONE 46.2s
```

we don't pin the mlflow version, instead we only set a lowerbound `mlflow>=2.3.1`, it looks like mlflow 2.7.0 can work with PyArrow 13.0.0:

`pyarrow [required: >=4.0.0,<14, installed: 13.0.0]`

**Comment by dongjoon-hyun** [comment]
Oh, it's great to have mlflow 2.7.0 on time! Thank you.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0. Thank you, @zhengruifeng and @HyukjinKwon .

**Comment by zhengruifeng** [comment]
thank you @dongjoon-hyun and @HyukjinKwon

**Pull request closed without merging**